src/app/backend/handler/localehandler.go

```diff
@@ -57,7 +50,7 @@ func CreateLocaleHandler() *LocaleHandler {
 		glog.Warningf("Error when loading the localization configuration. Dashboard will not be localized. %s", err)
 		locales = []language.Tag{}
 	}
-	return &LocaleHandler{SupportedLocales: locales}
+	return &LocaleHandler{SupportedLocales: locales, LocaleDirectories: getLocaleDirs()}
```

Comment:
I don't think we need `LocaleDirectories` at all. Old way of looking for matched locale was fine. Exact match + mapping and fallback to english if not found.

Reply:
To provide language service as much as we can, as you said too, this change provides base locale for Exact and High confidence that matched not strictly. Requirements makes logic complicated. It is unavoidable. Please see comments in source code. If we provide only strictly matched locales, we don't need to use confusing text/language module.

Reply:
If we no longer have `zh` support then high level match and falling back to base is obsolete. Mapping + exact match and fallback to english is enough right now.

Reply:
Agree about `zh`. I mean we can support following mappings.
* mapping `en-US` to `en`, `fr-FR` to `fr`, `ko-KR` to `ko`, etc in Exact match.
* mapping `en-GB` to `en`, `fr-CA` to `fr`, etc in High match.

Do you mean we don't need support them?

Reply:
We need to apply the same logic to other languages as to `zh`. The same principle applies to all of them.

Reply:
I think there are gaps between the principle, implementations in Angular, go module and browsers and user requirements.
As of now, carry over conclusions, prioritize the new release and wait for user feedbacks.

Reply:
In example, our `en` is actually `en-US`. There is no guideline as to how to handle base language and we usually have to decide on the region. Some countries have general locale, like i.e. Germany (`de`), but it has its' regional code also (`de-DE`).

In general, I'd like to avoid multiple indirect mappings due to all those subtle differences. It all makes handling localization more complex. We should decide and follow one strategy.

My idea would be to:
1. Map base language code `zh`, `de`, `fr` to the most common regional code. This is the only code that I'd be okay with to map in `angular.json` to a different directory. We'd have `zh-Hans`, `de-DE`, `fr-FR`. This would have to be noted clearly inside our documentation.
2. Always try to find an exact match first.
3. If exact match not found, return base.

This way even if base is returned and we do not support it, it will automatically fallback to English.

Reply:
Your idea 1 did not go well.
* Angular (https://github.com/angular/angular/tree/master/packages/common/locales) does not support `de-DE`, `fr-FR`, `ja-JP`, `ko-KR`. So warnings are shown while building locale for them. Angular seems to use `de` instead `de-DE`, etc. Only `zh*` locales seems odd.
* Also, in Angular docs https://angular.io/guide/i18n , new locales do not have region code for the most common region. e.g. `zh-Hans-CN` -> `zh-Hans`. `Hans` is not region.
* Many links take lots of memory while building binary. In my environment, it had taken over 16GB, so I could not finish the build.

Reply:
> Angular (https://github.com/angular/angular/tree/master/packages/common/locales) does not support de-DE, fr-FR, ja-JP, ko-KR. So warnings are shown while building locale for them. Angular seems to use de instead de-DE, etc. Only zh* locales seems odd.

I'll check. It should follow BCP47 standard and it allows regional codes.

> Also, in Angular docs https://angular.io/guide/i18n , new locales do not have region code for the most common region. e.g. zh-Hans-CN -> zh-Hans. Hans is not region.

True, `Hans` is actually a script code, not a region identifier. I guess Chinese is a bit more complicated than other languages. We still have to decide what to do here for the base locale.

> Many links take lots of memory while building binary. In my environment, it had taken over 16GB, so I could not finish the build.

With Angular 9 I don't believe it's true. Right now build is very fast and during the build, it did not use more than 3GB of RAM for me. It seems to be more CPU heavy.

Reply:
It seems that not supporting `en-US` and `de-DE` might be yet again a bug on angular side.
https://stackoverflow.com/questions/59514703/angular-9-locale-data-for-en-us-cannot-be-found-no-locale-data-will-be-includ

I'd really like to avoid so many workarounds, but it seems that there are issues on every step (angular, golang text library). :slightly_frowning_face:

Reply:
Current changes follow current Angular 9 spec. If we wait for consistency of all steps, we won't release new dashboard. For now, let's merge this and release new dashboard. 😉